### ISSUE TITLE
Unhandled Rejection (TypeError): Cannot read property 'refetch' of undefined

### ISSUE THREAD

**Issue by proIT324** [apollographql/react-apollo#3862]
Hello friends,

Today I faced an issue with GraphQL in my React.js project. What I tried to do is refetch a query once a mutation is completed. My codebase is as below

```javascript
const getUserTeamQuery = useQuery(getUserTeam(usersFragment))
const data = get(getUserTeamQuery, 'data.getUserTeam')

const [updateUser] = useMutation(updateUserMutation(usersFragment), {
  onCompleted: () => {
    getUserTeamQuery.refetch()
  }
})

const updateTeam = async () => {
  updateUser({ variables: { input: { pullTeam } } })
}
```

![Image error](https://i.ibb.co/pdpCMGD/Screen-Shot-2020-02-27-at-4-17-17-PM.png)

Is there any solution to fix the issue?

**Comment by jean9696** [comment]
Hi @proIT324, I just faced the same issue. It was because my component was unmount before the call of `refetch`. Maybe you have a side effect caused by your mutation which unmount the component were you call your `useQuery` ?

**Comment by EyMaddis** [comment]
I noticed the same thing while building a logout function that calles cache.reset(). Maybe this leads to an invalid state internally.

**Comment by alradadi** [comment]
Was running into the same issue, but noticed that it only appears after a **Live Reload**.
Works just fine after a full reload.

**Comment by forbesgillikin** [comment]
I am experiencing this as well in React Native when trying to refetch when the subscription client reconnects

```
SubscriptionClient.onReconnected(() => {
      refetch()
})
```

This works fine when the component loads initially. However, if I navigate to the previous screen, then back to the screen with the subscription `_this.currentObservable` is an empty object in 

```
_this.obsRefetch = function (variables) {
      return _this.currentObservable.query.refetch(variables);
};
```

thus, query is undefined.

**Comment by cryptedx** [comment]
I am experiencing the same problem. When I am loading my screen for the first time, my FlatList pull to refresh is working. After editing my screen.js file and saving it, right after that i will get this failure. I am also using chrome debug with fast refresh enabled. 

<img width="174" alt="error" src="https://user-images.githubusercontent.com/20707078/77852010-9254db80-71dc-11ea-8119-a95ff6cdc8fa.png">

<img width="470" alt="screenshot" src="https://user-images.githubusercontent.com/20707078/77851997-81a46580-71dc-11ea-8e3a-f0812d05d741.png">

**Comment by danieldunderfelt** [comment]
I am having the same problem in web React (Create React App and Typescript) with Fast Refresh. If a refetch happens after a fast refreshed change, it just crashes with this error. Would it be possible to fix it? It is probably not a production issue, but it is a huge annoyance when developing an app.

**Comment by cryptedx** [comment]
Indeed, it's so annoying. I always have to completely restart the app. Please please is a fix possible?

**Comment by vicary** [comment]
This also happens with Hermes, every time we save a file and triggers a refresh, refetching will throw.

**Comment by callaars** [comment]
The annoying thing is, this happens to us on production as well in combination with CodePush. Whenever a new version is updated the app will restart and voila, the whole app crashes because Apollo refetch is unavailable.

**Comment by cryptedx** [comment]
I thought it could be a problem because I had declared some anon functions. I fixed that but problem is still occurring.

**Comment by briangonzalez** [comment]
I current am working around this like so:

**Comment by briangonzalez** [comment]
I currently am working around this like so:

```
const { data, refetch } = useQuery(...) || {}
```

**Comment by oknoorap** [comment]
I'm using NextJS. I got the same error after upgrade to next `9.4`, happened after React refresh implemented by NextJS.

**Comment by d4vidES** [comment]
Use link not url in the configuration

`import { ApolloClient } from '@apollo/client';

import { InMemoryCache } from 'apollo-cache-inmemory';
import { HttpLink } from 'apollo-link-http';

const client = new ApolloClient({
    cache: new InMemoryCache({
        addTypename: true
    }),
    link: new HttpLink({
        uri: 'http://192.168.1.39:4000'
    })
});

export default client;`

**Comment by derekstavis** [comment]
I stumbled upon the same problem and it looks like  the culprit is the `!` operator in Apollo's code:

<img width="539" alt="image" src="https://user-images.githubusercontent.com/1611639/81968803-c5f39580-95d1-11ea-8daf-f164543385ab.png">

**Comment by vicary** [comment]
@derekstavis Is it some kind of pre-draft babel for optional chaining (`?.`)?

**Comment by callaars** [comment]
It's typescript, @vicary

**Comment by ethanIncentify** [comment]
^ https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator

**Comment by maxb94** [comment]
A quick fix is to wrap it in `setTimout` like so:

`const refetch = useCallback(() => setTimeout(() => _refetch(), 0), [_refetch])`

**Comment by maxb94** [comment]
A quick fix is to wrap it in `setTimout` like so:

`
const refetch = useCallback(() => {
    setTimeout(() => _refetch(), 0)
  }, [_refetch])
`

**Comment by vicary** [comment]
@callaars I was meant to be a joke but it pulled off poorly I guess, the use optional chaining would at least make it not crash.

**Comment by wooogler** [comment]
> h

That way works!
Thank you!

**Comment by wooogler** [comment]
> A quick fix is to wrap it in `setTimout` like so:
> 
> `const refetch = useCallback(() => { setTimeout(() => _refetch(), 0) }, [_refetch])`

const {loading, error, data} = useQuery(VIEW_ALL_COMMENTS);
const _refetch = useQuery(VIEW_ALL_COMMENTS).refetch;
const refetch = useCallback(() => { setTimeout(() => _refetch(), 0) }, [_refetch]);

This code works!
Thanks, maxb94!!

**Comment by tm1000** [comment]
@wooogler a better way to do that is this
```
const {loading, error, data, refetch: _refetch} = useQuery(VIEW_ALL_COMMENTS);
const refetch = useCallback(() => { setTimeout(() => _refetch(), 0) }, [_refetch]);
```

**Comment by tm1000** [comment]
FYI: https://github.com/apollographql/apollo-client/issues/5870

**Comment by vicary** [comment]
@tm1000 Thanks for taking my joke seriously, I guess. I have no idea about the inner workings, it was just a patchy-patch attitude. 😄

**Comment by peitalin** [comment]
Is this abandoned? or do we wait for v4.0 ?

**Comment by santaclauze** [comment]
^

**Comment by grantspeelman** [comment]
^^

**Comment by sgtpepper43** [comment]
^^^
